text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,416 @@
+    /// error is returned.
+    ///
+    /// # Panics
+    ///
+    /// If `f` panics, the panic is propagated to the caller, and the cell
+    /// remains uninitialized.
+    ///
+    /// It is an error to reentrantly initialize the cell from `f`. Doing
+    /// so results in a panic.
+    pub fn get_or_try_init<F, E>(&self, f: F) -> Result<&T, E>
+    where
+        F: FnOnce() -> Result<T, E>,
+    ;
+}
+```
+
+Notable features of the API:
+
+* `OnceCell` is created empty, by a const fn.
+* Initialization succeeds at most once.
+* `get_or_init` and `get_or_try_init` methods can be used to conveniently initialize a cell.
+* `get_` family of methods return `&T`.
+
+Similarly to other interior mutability primitives, `OnceCell` comes in two flavors:
+
+* Non thread-safe `std::cell::OnceCell`.
+* Thread-safe `std::sync::OnceCell`.
+
+Here's how `OnceCell` can be used to implement lazy-initialized global data:
+
+```rust
+use std::{sync::{Mutex, OnceCell}, collections::HashMap};
+
+fn global_data() -> &'static Mutex<HashMap<i32, String>> {
+    static INSTANCE: OnceCell<Mutex<HashMap<i32, String>>> = OnceCell::new();
+    INSTANCE.get_or_init(|| {
+        let mut m = HashMap::new();
+        m.insert(13, "Spica".to_string());
+        m.insert(74, "Hoyten".to_string());
+        Mutex::new(m)
+    })
+}
+```
+
+Here's how `OnceCell` can be used to implement a lazy field:
+
+```rust
+use std::{fs, io, path::PathBuf, cell::OnceCell};
+
+struct Ctx {
+    config_path: PathBuf,
+    config: OnceCell<String>,
+}
+
+impl Ctx {
+    pub fn get_config(&self) -> Result<&str, io::Error> {
+        let cfg = self.config.get_or_try_init(|| {
+            fs::read_to_string(&self.config_path)
+        })?;
+        Ok(cfg.as_str())
+    }
+}
+```
+
+We also provide a more convenient but less powerful `Lazy<T, F>` wrapper around `OnceCell<T>`, which allows to specify the initializing closure at creation time:
+
+```rust
+pub struct Lazy<T, F = fn() -> T> { ... }
+
+impl<T, F> Lazy<T, F> {
+    /// Creates a new lazy value with the given initializing function.
+    pub const fn new(init: F) -> Lazy<T, F>;
+}
+
+impl<T, F: FnOnce() -> T> Lazy<T, F> {
+    /// Forces the evaluation of this lazy value and returns a reference to
+    /// the result.
+    ///
+    /// This is equivalent to the `Deref` impl, but is explicit.
+    pub fn force(this: &Lazy<T, F>) -> &T;
+}
+
+impl<T, F: FnOnce() -> T> Deref for Lazy<T, F> {
+    type Target = T;
+
+    fn deref(&self) -> &T;
+}
+```
+
+`Lazy` directly replaces `lazy_static!`:
+
+```rust
+use std::{sync::{Mutex, Lazy}, collections::HashMap};
+
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>> = Lazy::new(|| {
+    let mut m = HashMap::new();
+    m.insert(13, "Spica".to_string());
+    m.insert(74, "Hoyten".to_string());
+    Mutex::new(m)
+});
+```
+
+Moreover, once `#[thread_local]` attribute is stable, `Lazy` will supplant `std::thread_local!` as well:
```

Comment:
`thread_local!` actually does a lot more than `lazy_static!` since it needs to set up destructors to run when the thread exits. It also needs to ensure the thread-local data isn't accessed after it has been dropped.

Reply:
Hm, indeed, I for some reason assumed that destructor management was handled by the compiler, but [this is done by the library](https://github.com/rust-lang/rust/blob/b7a9c285a50f3a94c44687ba9ff3ab0648243aaa/src/libstd/thread/local.rs#L428-L470).